### ISSUE TITLE
init regression in 0.3.0: falsely detected Thumb function start due to data that looks like a shift left by 4 instruction

### ISSUE THREAD

**Issue by LagoLunatic** [AetiasHax/ds-decomp#22]
For the Japanese version of Castlevania: Order of Ecclesia (YR9J00), 0.3.0's analysis detects two enormous false positive Thumb functions in a specific overlay, which were previously correctly detected as data:
![image](https://github.com/user-attachments/assets/017fc8e7-90ae-4f74-abda-c1d8b75009f3)

(The US version isn't affected, because even though the data is almost the same it just so happens that one of the addresses further down in this huge chunk of data looks like an illegal Thumb instruction which aborts it for US but not JP.)

I was curious what changed so I bisected it to this commit: e7eb2d0c230184ae112c86a52bb174d4aa2dd362
The commit message says "Allow function to start with shift by 16 or 24", but the weird part is neither of the two false functions it detects start like that.
0x022c3d80 and 0x022c485c look like this (after forcing ghidra to disassemble the data as Thumb):
![image](https://github.com/user-attachments/assets/ae31a8ec-c638-4bb0-aca9-c469e59a6f90)
![image](https://github.com/user-attachments/assets/455d1233-bc20-4226-94c5-f6a4a1ff73d4)

Is `lsl r2, r0, 0x4` being an allowed start a bug? Or do some games have functions that start like that?

**Comment by AetiasHax** [comment]
> Is `lsl r2, r0, 0x4` being an allowed start a bug? Or do some games have functions that start like that?

I suppose it's not a bug, the heuristic in the mentioned commit only applies when the two registers are the same: `src == dst`. I tried removing this constraint, which didn't affect any of my roundtrip tested games, so I guess it's safe to remove the `src == dst` altogether. Let me know if that works and I'll push it to main afterwards.

(I'm not happy about how arbitrary this heuristic is. Hopefully it can be completely removed in the future.)

**Comment by LagoLunatic** [comment]
Just tried out removing `src == dst`. It fixes the 0.3 regression mentioned in the OP, which applies to YR9J and YR9P.

While I was testing YR9P, I noticed that this change also fixes a second instance of a huge false thumb function in a different overlay for that region (`func_ov083_022c2760`) which I wasn't aware of until now because this one was also falsely detected by 0.2.

Other than those two, it didn't affect the analysis of any of my other games that roundtrip currently runs without errors for.

It also fixes #23 - all four regions of PoR now roundtrip test successfully and analysis seems more accurate. Which is kind of unexpected, since that was caused by a different commit than this one. But I'll take it.
